Public/Src/Cache/ContentStore/Library/Stores/FileSystemContentStoreInternal.cs

```
@@ -349,6 +354,195 @@ private Task Upgrade2To3(Context context)
+            await EvictAsync(
+                context,
+                new ContentHashWithLastAccessTimeAndReplicaCount(contentHash, Clock.UtcNow),
+                onlyUnlinked: false,
+                size => { QuotaKeeper.OnContentEvicted(size); })
+                .TraceIfFailure(context);
+
+            if (_distributedEvictionSettings?.DistributedStore != null)
+            {
+                await _distributedEvictionSettings.DistributedStore.UnregisterAsync(context, new ContentHash[] {contentHash}, context.Token)
+                    .TraceIfFailure(context);
+            }
+        }
+
+        private async Task<(bool isValid, string error)> ValidateFileAsync(Context context, ContentHash expectedHash, FileInfo fileInfo)
+        {
+            try
+            {
+                // The cache entry is invalid if the size in content directory doesn't mach an actual size
+                if (ContentDirectory.TryGetFileInfo(expectedHash, out var contentFileInfo) && contentFileInfo.FileSize != fileInfo.Length)
+                {
+                    return (isValid: false, error: $"File size mismatch. Expected size is {contentFileInfo.FileSize} and size on disk is {fileInfo.Length}.");
+                }
+
+                // Or if the content doesn't match the hash.
+                var actualHashAndSize = await TryHashFileAsync(context, fileInfo.FullPath, expectedHash.HashType);
+                if (actualHashAndSize != null && actualHashAndSize.Value.Hash != expectedHash)
+                {
+                    // Don't need to add an expected hash into the error string because the client code will always put it into the final error message.
+                    return (isValid: false, error: $"Hash mismatch. Actual hash is {actualHashAndSize.Value.Hash}.");
+                }
+
+                return (isValid: true, error: string.Empty);
+            }
+            catch (IOException e)

```

>IOException e [](start = 19, length = 13)

Catch all exceptions